src_features/signMessage/cmd_signMessage.c

```diff
@@ -1,199 +1,292 @@
 #include <stdbool.h>
-#include "shared_context.h"
+#include <ctype.h>
+#include <string.h>
 #include "apdu_constants.h"
-#include "utils.h"
-#include "common_ui.h"
+#include "sign_message.h"
+#include "ui_flow_signMessage.h"
+
+static uint8_t processed_size;
+static struct {
+    sign_message_state sign_state : 1;
+    bool ui_started : 1;
+} states;
 
 static const char SIGN_MAGIC[] =
     "\x19"
     "Ethereum Signed Message:\n";
 
 /**
- * Check if a given character is a "special" displayable ASCII character
+ * Send a response APDU with the given Status Word
  *
- * @param[in] c character we're checking
- * @return wether the character is special or not
+ * @param[in] sw status word
  */
-static inline bool is_char_special(char c) {
-    return ((c >= '\b') && (c <= '\r'));
+static void apdu_reply(uint16_t sw) {
+    *(uint16_t *) G_io_apdu_buffer = __builtin_bswap16(sw);
+    io_exchange(CHANNEL_APDU | IO_RETURN_AFTER_TX, 2);
 }
 
 /**
- * Check if a given data is made of ASCII characters
+ * Get unprocessed data from last received APDU
  *
- * @param[in] data the input data
- * @param[in] the length of the input data
- * @return wether the data is fully ASCII or not
+ * @return pointer to data in APDU buffer
  */
-static bool is_data_ascii(const uint8_t *const data, size_t length) {
-    for (uint8_t idx = 0; idx < length; ++idx) {
-        if (!is_char_special(data[idx]) && ((data[idx] < 0x20) || (data[idx] > 0x7e))) {
-            return false;
-        }
-    }
-    return true;
+static const uint8_t *unprocessed_data(void) {
+    return &G_io_apdu_buffer[OFFSET_CDATA] + processed_size;
 }
 
 /**
- * Initialize value string that will be displayed in the UX STEP
+ * Get size of unprocessed data from last received APDU
  *
- * @param[in] if the value is ASCII
+ * @return size of data in bytes
  */
-static void init_value_str(bool is_ascii) {
-    if (is_ascii) {
-        strings.tmp.tmp[0] = '\0';  // init string as empty
-    } else {
-        strcpy(strings.tmp.tmp, "0x");  // will display the hex bytes instead
-    }
+static size_t unprocessed_length(void) {
+    return G_io_apdu_buffer[OFFSET_LC] - processed_size;
```

Comment:
The data length in `G_io_apdu_buffer[OFFSET_LC]` could be a lie and not match the actual length of the data. It's kind of ok here since we're displaying everything and the user would see garbage from the previous APDU, but it can be fixed easily and for all APDUs I think, by always checking that the `rx` value returned from `io_exchange` is `G_io_apdu_buffer[OFFSET_LC] + 5`. What do you think?

Reply:
I'm confused that this is not already checked in `main:app_main`.